pkg/component/networking/istio/charts/istio/istio-ingress/templates/apiserver-tls-termination-envoyfilter.yaml

```diff
@@ -0,0 +1,79 @@
+{{- if eq .Values.terminateLoadBalancerAPIServer true }}
+apiVersion: networking.istio.io/v1alpha3
+kind: EnvoyFilter
+metadata:
+  labels:
+{{ .Values.labels | toYaml | indent 4 }}
+  name: apiserver-tls-termination
+  namespace: {{ .Release.Namespace }}
+spec:
+  workloadSelector:
+    labels:
+{{ .Values.labels | toYaml | indent 6 }}
+  configPatches:
+  - applyTo: HTTP_FILTER
+    match:
+      context: GATEWAY
+      listener:
+        portNumber: 9443
+        filterChain:
+          filter:
+            name: "envoy.filters.network.http_connection_manager"
+            subFilter:
+              name: "envoy.filters.http.router"
+    patch:
+      operation: INSERT_BEFORE
+      value:
+        name: envoy.filters.http.lua
+        typed_config:
+          "@type": "type.googleapis.com/envoy.extensions.filters.http.lua.v3.Lua"
+          defaultSourceCode:
+            inlineString: |
+              local function startsWith(str, prefix)
+                return string.sub(str, 1, #prefix) == prefix
+              end
+
+              function envoy_on_request(request_handle)
+                -- Drop all x-remote- headers.
+                local remove = {}
+                for key, value in pairs(request_handle:headers()) do
+                  if startsWith(key:lower(), "x-remote-") then
+                    remove[key] = ""
+                  end
+                end
+                for key, value in pairs(remove) do
+                  request_handle:headers():remove(key)
+                end
+
+                local streamInfo = request_handle:streamInfo()
+                local ssl = streamInfo:downstreamSslConnection()
+
+                -- TODO(oliver-goetz) switch to ssl:parsedSubjectPeerCertificate() once it is available.
+                local cn = ssl:subjectPeerCertificate()
+                if cn ~= "" then
+                  -- Get the host header from the request and set it as dynamic metadata.
+                  local host = request_handle:headers():get(":authority")
+                  if not host then
+                    host = request_handle:headers():get("host")
+                  end
+                  streamInfo:dynamicMetadata():set("envoy.filters.http.lua", "{{ .Values.apiServerAuthenticationDynamicMetadataKey }}", host)
+
+                  -- Add request headers for kube-apiserver authentication.
+                  local remoteUserHeaderAdded = false
+                  for pair in string.gmatch(cn, "([^,]+)") do
+                      local key, value = pair:match("([^=]+)=([^=]+)")
+                      if key == "CN" then
+                        request_handle:headers():add("{{ .Values.apiServerRequestHeaderUserName }}", value)
+                        remoteUserHeaderAdded = true
+                      elseif key == "O" then
+                        request_handle:headers():add("{{ .Values.apiServerRequestHeaderGroup }}", value)
+                      end
+                  end
```

Comment:
What is going on here?
I assume that we take a peek into the client certificate and try to find the common name. In case of token-based authentication (or no authentication) we would probably skip the `if` block.
Why do we look at the http host authority/header instead of the SNI name? The latter one seems to be better suited. Otherwise, you might open interesting attack vectors.

Please add more comments especially regarding the pattern matching. I do not get it.

Reply:
`if cn ~= ""` in lua means `if cn != ""` in many other programming languages. Thus, yes the block is skipped if there is no client certificate.

We get the `host` from SNI now and I added a couple of comments.